lib/src/piece/case.dart

```diff
@@ -54,17 +61,23 @@ final class CaseExpressionPiece extends Piece {
 
   @override
   List<State> get additionalStates => [
+        if (_canBlockSplitBody) _blockSplitBody,
         _beforeBody,
         if (_guard != null) ...[_beforeWhenAndBody],
       ];
 
   @override
   bool allowNewlineInChild(State state, Piece child) {
     return switch (state) {
+      // If the outermost pattern is `||`, then always let it split even while
+      // allowing the body on the same line as `=>`.
+      _ when child == _pattern && _patternIsLogicalOr => true,
```

Comment:
Why do we do this for `||` but not `&&`?

Reply:
The intent of this formatting is to treat:

```dart
switch (obj) {
  constant ||
  another => body,
};
```

Sort of like:

```dart
switch (obj) {
  case constant:
  case another:
    body;
};
```

Since switch expressions don't allow multiple cases to share a body, using `||` as the top level pattern emulates that and yields similar formatting where each arm is as the beginning of the line (assuming they need to split).

In other words, readers seeing a `||` pattern at the top level of an expression switch case read it like a series of independent cases with a single body, and we format it like that.

There's no similar analogue for `&&`. In that pattern, you need to reason about all of the arms together to understand when the pattern matches.